src/sincosint.jl

```diff
@@ -0,0 +1,258 @@
+const pidiv2 = π/2
+
+import Base.Math: @horner
+
+# Compute the sine integral: ∫_0^x sin(t)/t dt,
+# and the cosine integral: γ + log x + ∫_0^x (cos(t)-1)/t dt,
+# using the rational approximants tabulated in:
+#     A.J. MacLeod, "Rational approximations, software and test methods for
+#     sine and cosine integrals," Numer. Algor. 12, pp. 259--272 (1996).
+#         http://dx.doi.org/10.1007/BF02142806
+#         https://link.springer.com/article/10.1007/BF02142806
+#
+# Note: the second zero of Ci(x) has a typo that is fixed:
+#
+# r1 = 3.38418 04228 51186 42639 78511 46402 in the article, but is in fact:
+#
+# r1 = 3.38418 04225 51186 42639 78511 46402.
+#
+
+function sinint(x::Float64)
+    t = x*x
+    if t ≤ 36.0
+        return x * @horner(t, 1.00000000000000000000E0,
+                             -0.44663998931312457298E-1,
+                              0.11209146443112369449E-2,
+                             -0.13276124407928422367E-4,
+                              0.85118014179823463879E-7,
+                             -0.29989314303147656479E-9,
+                              0.55401971660186204711E-12,
+                             -0.42406353433133212926E-15) /
+                   @horner(t, 1.00000000000000000000E0,
+                              0.10891556624243098264E-1,
+                              0.59334456769186835896E-4,
+                              0.21231112954641805908E-6,
+                              0.54747121846510390750E-9,
+                              0.10378561511331814674E-11,
+                              0.13754880327250272679E-14,
+                              0.10223981202236205703E-17)
+    elseif t ≤ 144.0
+        invt = inv(t)
```

Comment:
@stevengj, your benchmark shows that Julia's implementation of complex numbers can be improved. I get the same performance from `inv(3.0+4.0im)` as `1.0/(3.0+4.0im)` *if I remove the unnecessary type conversion & promotion in `/`:

```julia
julia> import Base: /

julia> /(a::Float64, z::Complex128) = a*inv(z)
/ (generic function with 74 methods)

julia> z = 3.0+4.0im
3.0 + 4.0im

julia> h(z) = 1.0/z
h (generic function with 1 method)

julia> @btime h($z)
  9.052 ns (0 allocations: 0 bytes)
0.12 - 0.16im

julia> @btime inv($z)
  9.052 ns (0 allocations: 0 bytes)
0.12 - 0.16im
```

The following is even faster:
```julia
julia> /(a::Float64, z::Complex128) = (ainvabs2z = a*inv(abs2(z)); Complex(real(z)*ainvabs2z, -imag(z)*ainvabs2z))
/ (generic function with 74 methods)

julia> h(z) = 1.0/z
h (generic function with 1 method)

julia> @btime h($z)
  3.621 ns (0 allocations: 0 bytes)
0.12 - 0.16im

```
but it doesn't use Base's specialized `inv(z::Complex128)`, so it's not clear that the output should be the same for all input.

So, in my opinion, the moral of the story is that the performance of `inv(t)`, `1/t`, `one(t)/t`, `1.0/t`, `one(eltype(t))/t` ... depends on the *context and implementation*.

In the present context that `t::Float64`, all methods appear to be implemented with equivalent performance.